api/apps/geoprocessing/src/import/pieces-importers/scenario-protected-areas.piece-importer.ts

```diff
@@ -0,0 +1,143 @@
+import { extractFile } from '@marxan/utils';
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectEntityManager } from '@nestjs/typeorm';
+import { isLeft } from 'fp-ts/lib/Either';
+import { EntityManager } from 'typeorm';
+import {
+  ImportPieceProcessor,
+  PieceImportProvider,
+} from '../pieces/import-piece-processor';
+
+@Injectable()
+@PieceImportProvider()
+export class ScenarioProtectedAreasPieceImporter
+  implements ImportPieceProcessor {
+  constructor(
+    private readonly fileRepository: FileRepository,
+    @InjectEntityManager(geoprocessingConnections.apiDB)
+    private readonly apiEntityManager: EntityManager,
+    @InjectEntityManager(geoprocessingConnections.default)
+    private readonly geoprocessingEntityManager: EntityManager,
+    private readonly logger: Logger,
+  ) {
+    this.logger.setContext(ScenarioProtectedAreasPieceImporter.name);
+  }
+
+  isSupported(piece: ClonePiece): boolean {
+    return piece === ClonePiece.ScenarioProtectedAreas;
+  }
+
+  async run(input: ImportJobInput): Promise<ImportJobOutput> {
+    const {
+      importResourceId: projectId,
+      componentResourceId: scenarioId,
+      uris,
+      piece,
+    } = input;
+
+    if (uris.length !== 1) {
+      const errorMessage = `uris array has an unexpected amount of elements: ${uris.length}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+    const [scenarioProtectedAreasLocation] = uris;
+
+    const readableOrError = await this.fileRepository.get(
+      scenarioProtectedAreasLocation.uri,
+    );
+
+    if (isLeft(readableOrError)) {
+      const errorMessage = `File with piece data for ${piece}/${scenarioId} is not available at ${scenarioProtectedAreasLocation.uri}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+
+    const stringScenarioProtectedAreasOrError = await extractFile(
+      readableOrError.right,
+      scenarioProtectedAreasLocation.relativePath,
+    );
+    if (isLeft(stringScenarioProtectedAreasOrError)) {
+      const errorMessage = `Scenario protected areas file extraction failed: ${scenarioProtectedAreasLocation.relativePath}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+
+    const {
+      threshold,
+      wdpa,
+      customProtectedAreas,
+    }: ScenarioProtectedAreasContent = JSON.parse(
+      stringScenarioProtectedAreasOrError.right,
+    );
+    const protectedAreasIds: string[] = [];
+
+    const wdpaProtectedAreas: {
+      id: string;
+      wdpaid: number;
+    }[] = await this.geoprocessingEntityManager
+      .createQueryBuilder()
+      .select('id')
+      .addSelect('wdpaid')
+      .from(ProtectedArea, 'pa')
+      .where('wdpaid IN (:...wdpaIds)', { wdpaIds: wdpa })
+      .execute();
+
+    if (wdpaProtectedAreas.length !== wdpa.length) {
+      const notFoundIds = wdpa.filter(
+        (id) => !wdpaProtectedAreas.map((pa) => pa.wdpaid).includes(id),
```

Comment:
The map for `wdpaid` should be done outside the filter callback to be optimal. Due to this array size is not a big deal tho.